core/src/main/java/org/apache/struts2/interceptor/ExecuteAndWaitInterceptor.java

```diff
@@ -394,7 +404,10 @@ public void init() {
 
     @Override
     public void destroy() {
-        super.destroy();
-        executor.shutdown();
+        try {
+          executor.shutdown();
+        } finally {
+          super.destroy();
+        }
```

Comment:
👏

Reply:
Hi Lukasz.  Thanks for reviewing the proposed fix changes.  😄 

Looking at the commit history, `SessionMap`  has extended `AbstractMap` since its earliest commits (many, many years ago).  Since `SessionMap` is _public_ and has been structured a certain way for so long, changing its design to delegate (to a non-accessible internal map, instead of extending) could have unexpected impacts (especially for anything that might try to use/extend it in user-space code).

Staying with the long-standing (preexisting) design might be safer in terms of potential side-effects.  Keeping `@overrides` annotations in place along with some unit tests that attempt to detect breaking-changes should help.  Considering a `SessionMap` redesign would probably need to be a standalone issue/work item, if people think it is needed (taking into account the existing design seems to have worked sufficiently for a long time).

Hopefully my reasoning above makes sense.  If it seems like other changes/improvements should be made to the PR, please let me know.

Reply:
Thanks a lot for your comment, yet I was thinking about redesigning the current *Map approach (there is ApplicationMap, RequestMap as well) and use native objects to have more control over them. Right now we must prevent extra cuotion to secure the framweork (excluded `#application` or `#request` in ParametersInterceptor).

Anyway this will happen in Struts 7.0 at leas :)